File: funtime/lib/snake-draw.js
/* ============================================================
   snake-draw.js — everything you can see

   Black and white only, like the very first handheld snake games: the field is
   white, the snake is black, and the apple is a black circle.

   Every function here takes a canvas "ctx" (the drawing tool) and paints
   something. None of them change the game — drawing never cheats.
   ============================================================ */

const COLOR_BACKGROUND = '#ffffff';
const COLOR_GRID = '#ececec';
const COLOR_SNAKE = '#111111';
const COLOR_INNER = '#ffffff';
const COLOR_FRAME = '#111111';

/**
 * clearCanvas — paint the whole canvas one flat colour.
 * INPUT: ctx, width, height, color. OUTPUT: nothing.
 * ALGORITHM: set fillStyle, then fillRect over everything.
 */
function clearCanvas(ctx, width, height, color) {
    ctx.fillStyle = color;
    ctx.fillRect(0, 0, width, height);
}

/**
 * drawGrid — the faint lines that show the squares.
 * INPUT: ctx, columns, rows, cellSize. OUTPUT: nothing.
 * ALGORITHM: one vertical line per column, one horizontal line per row.
 */
function drawGrid(ctx, columns, rows, cellSize) {
    ctx.strokeStyle = COLOR_GRID;
    ctx.lineWidth = 1;
    for (let c = 0; c <= columns; c++) {
        ctx.beginPath();
        ctx.moveTo(c * cellSize + 0.5, 0);
        ctx.lineTo(c * cellSize + 0.5, rows * cellSize);
        ctx.stroke();
    }
    for (let r = 0; r <= rows; r++) {
        ctx.beginPath();
        ctx.moveTo(0, r * cellSize + 0.5);
        ctx.lineTo(columns * cellSize, r * cellSize + 0.5);
        ctx.stroke();
    }
}

/**
 * drawSegment — one square of the snake's body.
 *
 * INPUT:  ctx, position, cellSize
 * OUTPUT: nothing
 *
 * ALGORITHM: a black square with a smaller white square outlined inside it, so
 *            you can count the segments even when the snake is packed tight.
 */
function drawSegment(ctx, position, cellSize) {
    const x = position.x * cellSize;
    const y = position.y * cellSize;

    ctx.fillStyle = COLOR_SNAKE;
    ctx.fillRect(x + 1, y + 1, cellSize - 2, cellSize - 2);

    /* The hollow middle is sized from the square, so the snake still reads
       clearly on the small board the workshop uses. */
    const inset = Math.max(3, Math.round(cellSize * 0.26));
    ctx.strokeStyle = COLOR_INNER;
    ctx.lineWidth = Math.max(1, Math.round(cellSize * 0.1));
    ctx.strokeRect(x + inset, y + inset, cellSize - inset * 2, cellSize - inset * 2);
}

/**
 * drawHead — the snake's head, with two eyes looking where it is going.
 *
 * INPUT:  ctx, position, direction, cellSize
 * OUTPUT: nothing
 *
 * ALGORITHM:
 *   1. Fill the whole square black (no hollow middle — that is what makes the
 *      head stand out from the body).
 *   2. Find the middle of the square.
 *   3. Step a little way along the direction to reach the face, then a little
 *      way to each side (the sideways step of (x, y) is (-y, x)) — and put a
 *      small white eye at each of those two spots.
 */
function drawHead(ctx, position, direction, cellSize) {
    const x = position.x * cellSize;
    const y = position.y * cellSize;

    ctx.fillStyle = COLOR_SNAKE;
    ctx.fillRect(x + 1, y + 1, cellSize - 2, cellSize - 2);

    const middleX = x + cellSize / 2;
    const middleY = y + cellSize / 2;
    const forward = cellSize * 0.20;
    const sideways = cellSize * 0.22;
    const eyeSize = Math.max(2, Math.round(cellSize * 0.13));

    const faceX = middleX + direction.x * forward;
    const faceY = middleY + direction.y * forward;
    const sideX = -direction.y * sideways;
    const sideY = direction.x * sideways;

    ctx.fillStyle = COLOR_INNER;
    ctx.fillRect(faceX + sideX - eyeSize / 2, faceY + sideY - eyeSize / 2, eyeSize, eyeSize);
    ctx.fillRect(faceX - sideX - eyeSize / 2, faceY - sideY - eyeSize / 2, eyeSize, eyeSize);
}

/**
 * drawFood — the apple.
 *
 * INPUT:  ctx, position, cellSize
 * OUTPUT: nothing
 *
 * ALGORITHM: a filled black circle in the middle of the square, with a small
 *            white dot on it so it never looks like a body segment.
 */
function drawFood(ctx, position, cellSize) {
    const middleX = position.x * cellSize + cellSize / 2;
    const middleY = position.y * cellSize + cellSize / 2;

    ctx.fillStyle = COLOR_SNAKE;
    ctx.beginPath();
    ctx.arc(middleX, middleY, cellSize * 0.38, 0, Math.PI * 2);
    ctx.fill();

    ctx.fillStyle = COLOR_INNER;
    ctx.beginPath();
    ctx.arc(middleX - cellSize * 0.12, middleY - cellSize * 0.12, cellSize * 0.09, 0, Math.PI * 2);
    ctx.fill();
}

/**
 * drawFrame — the thick black border around the field.
 * INPUT: ctx, width, height. OUTPUT: nothing.
 * ALGORITHM: stroke a rectangle just inside the edge of the canvas.
 */
function drawFrame(ctx, width, height) {
    ctx.strokeStyle = COLOR_FRAME;
    ctx.lineWidth = 3;
    ctx.strokeRect(1.5, 1.5, width - 3, height - 3);
}

/**
 * drawMessage — big centred words across the field.
 *
 * INPUT:  ctx, width, height, title, subtitle
 * OUTPUT: nothing
 *
 * ALGORITHM: cover the field with a see-through white sheet, then write the
 *            title and the subtitle in the middle. The letters are sized from
 *            the width of the canvas so they still fit on a small board.
 */
function drawMessage(ctx, width, height, title, subtitle) {
    ctx.fillStyle = 'rgba(255, 255, 255, 0.88)';
    ctx.fillRect(0, 0, width, height);

    ctx.fillStyle = COLOR_SNAKE;
    ctx.textAlign = 'center';

    const scale = width / 400;
    const titleSize = Math.max(14, Math.round(34 * scale));
    const subtitleSize = Math.max(9, Math.round(15 * scale));

    ctx.font = 'bold ' + titleSize + 'px monospace';
    ctx.fillText(title, width / 2, height / 2 - 8);

    ctx.font = subtitleSize + 'px monospace';
    ctx.fillText(subtitle, width / 2, height / 2 + titleSize * 0.8);
}

/**
 * renderGame — draw one complete frame.
 *
 * INPUT:  ctx — the drawing tool. state — the game state. cellSize — pixels per square.
 * OUTPUT: nothing
 *
 * ALGORITHM (order matters — later things cover earlier ones):
 *   1. Clear to white and draw the faint grid.
 *   2. Draw the apple.
 *   3. Draw every body segment, then the head on top.
 *   4. Draw the black frame.
 *   5. If the game is won, lost or paused, write the message over everything.
 */
function renderGame(ctx, state, cellSize) {
    const width = GRID_WIDTH * cellSize;
    const height = GRID_HEIGHT * cellSize;

    clearCanvas(ctx, width, height, COLOR_BACKGROUND);
    drawGrid(ctx, GRID_WIDTH, GRID_HEIGHT, cellSize);

    if (state.food !== null) {
        drawFood(ctx, state.food, cellSize);
    }

    for (let i = 1; i < state.snake.length; i++) {
        drawSegment(ctx, state.snake[i], cellSize);
    }
    if (state.snake.length > 0) {
        drawHead(ctx, state.snake[0], state.direction, cellSize);
    }

    drawFrame(ctx, width, height);

    if (state.isWon) {
        drawMessage(ctx, width, height, 'YOU WIN!', 'You filled the whole board!');
    } else if (state.isOver) {
        drawMessage(ctx, width, height, 'GAME OVER', 'Press R or tap New to play again');
    } else if (state.isPaused) {
        drawMessage(ctx, width, height, 'PAUSED', 'Press P or tap Play to start');
    }
}

File: funtime/pylib/snake_draw.py
"""snake_draw.py - drawing the Snake game, in Python.

Yes: this Python code paints on an HTML canvas. Pyodide lets Python call the
same drawing tools JavaScript uses, so `ctx.fillRect(...)` works here exactly
as it does there.

Every function takes the canvas drawing tool `ctx` as its first input. None of
them change the game - drawing never cheats.
"""

from snake_rules import GRID_WIDTH, GRID_HEIGHT

COLOR_BACKGROUND = "#ffffff"
COLOR_GRID = "#ececec"
COLOR_SNAKE = "#111111"
COLOR_INNER = "#ffffff"


def clear_canvas(ctx, width, height, color):
    """Paint the whole canvas one flat colour.

    INPUT:  ctx, width, height, color. OUTPUT: nothing.
    ALGORITHM: set fillStyle, then fillRect over everything.
    """
    ctx.fillStyle = color
    ctx.fillRect(0, 0, width, height)


def draw_grid(ctx, columns, rows, cell_size):
    """The faint lines that show the squares.

    INPUT:  ctx, columns, rows, cell_size. OUTPUT: nothing.
    ALGORITHM: one line per column, then one per row.
    """
    ctx.strokeStyle = COLOR_GRID
    ctx.lineWidth = 1
    for c in range(columns + 1):
        ctx.beginPath()
        ctx.moveTo(c * cell_size + 0.5, 0)
        ctx.lineTo(c * cell_size + 0.5, rows * cell_size)
        ctx.stroke()
    for r in range(rows + 1):
        ctx.beginPath()
        ctx.moveTo(0, r * cell_size + 0.5)
        ctx.lineTo(columns * cell_size, r * cell_size + 0.5)
        ctx.stroke()


def draw_segment(ctx, position, cell_size):
    """One square of the snake's body.

    INPUT:  ctx, position, cell_size. OUTPUT: nothing.
    ALGORITHM: a black square with a smaller white square outlined inside it.
    """
    x = position[0] * cell_size
    y = position[1] * cell_size

    ctx.fillStyle = COLOR_SNAKE
    ctx.fillRect(x + 1, y + 1, cell_size - 2, cell_size - 2)

    inset = max(3, round(cell_size * 0.26))
    ctx.strokeStyle = COLOR_INNER
    ctx.lineWidth = max(1, round(cell_size * 0.1))
    ctx.strokeRect(x + inset, y + inset, cell_size - inset * 2, cell_size - inset * 2)


def draw_head(ctx, position, direction, cell_size):
    """The snake's head, with two eyes looking where it is going.

    INPUT:  ctx, position, direction, cell_size. OUTPUT: nothing.
    ALGORITHM: fill the square solid (no hollow middle, so the head stands out),
    then step forward from the middle to the face and out to each side - the
    sideways step of (x, y) is (-y, x) - and put a white eye at each spot.
    """
    x = position[0] * cell_size
    y = position[1] * cell_size

    ctx.fillStyle = COLOR_SNAKE
    ctx.fillRect(x + 1, y + 1, cell_size - 2, cell_size - 2)

    middle_x = x + cell_size / 2
    middle_y = y + cell_size / 2
    forward = cell_size * 0.20
    sideways = cell_size * 0.22
    eye = max(2, round(cell_size * 0.13))

    face_x = middle_x + direction[0] * forward
    face_y = middle_y + direction[1] * forward
    side_x = -direction[1] * sideways
    side_y = direction[0] * sideways

    ctx.fillStyle = COLOR_INNER
    ctx.fillRect(face_x + side_x - eye / 2, face_y + side_y - eye / 2, eye, eye)
    ctx.fillRect(face_x - side_x - eye / 2, face_y - side_y - eye / 2, eye, eye)


def draw_food(ctx, position, cell_size):
    """The apple.

    INPUT:  ctx, position, cell_size. OUTPUT: nothing.
    ALGORITHM: a filled black circle with a small white dot for shine.
    """
    import math

    middle_x = position[0] * cell_size + cell_size / 2
    middle_y = position[1] * cell_size + cell_size / 2

    ctx.fillStyle = COLOR_SNAKE
    ctx.beginPath()
    ctx.arc(middle_x, middle_y, cell_size * 0.38, 0, math.pi * 2)
    ctx.fill()

    ctx.fillStyle = COLOR_INNER
    ctx.beginPath()
    ctx.arc(middle_x - cell_size * 0.12, middle_y - cell_size * 0.12, cell_size * 0.09, 0, math.pi * 2)
    ctx.fill()


def draw_frame(ctx, width, height):
    """The thick black border around the field.

    INPUT:  ctx, width, height. OUTPUT: nothing.
    """
    ctx.strokeStyle = COLOR_SNAKE
    ctx.lineWidth = 3
    ctx.strokeRect(1.5, 1.5, width - 3, height - 3)


def draw_message(ctx, width, height, title, subtitle):
    """Big centred words across the field.

    INPUT:  ctx, width, height, title, subtitle. OUTPUT: nothing.
    ALGORITHM: cover the field with a see-through white sheet, then write the
    two lines, sized from the width of the canvas.
    """
    ctx.fillStyle = "rgba(255, 255, 255, 0.88)"
    ctx.fillRect(0, 0, width, height)

    ctx.fillStyle = COLOR_SNAKE
    ctx.textAlign = "center"

    scale = width / 400
    title_size = max(14, round(34 * scale))
    subtitle_size = max(9, round(15 * scale))

    ctx.font = "bold " + str(title_size) + "px monospace"
    ctx.fillText(title, width / 2, height / 2 - 8)

    ctx.font = str(subtitle_size) + "px monospace"
    ctx.fillText(subtitle, width / 2, height / 2 + title_size * 0.8)


def render_game(ctx, state, cell_size):
    """Draw one complete frame.

    INPUT:  ctx, state, cell_size. OUTPUT: nothing.
    ALGORITHM (order matters - later things cover earlier ones):
      1. Clear to white and draw the faint grid.
      2. Draw the apple.
      3. Draw the body, then the head on top.
      4. Draw the border, and any message.
    """
    width = GRID_WIDTH * cell_size
    height = GRID_HEIGHT * cell_size

    clear_canvas(ctx, width, height, COLOR_BACKGROUND)
    draw_grid(ctx, GRID_WIDTH, GRID_HEIGHT, cell_size)

    if state["food"] is not None:
        draw_food(ctx, state["food"], cell_size)

    for segment in state["snake"][1:]:
        draw_segment(ctx, segment, cell_size)
    if state["snake"]:
        draw_head(ctx, state["snake"][0], state["direction"], cell_size)

    draw_frame(ctx, width, height)

    if state["is_won"]:
        draw_message(ctx, width, height, "YOU WIN!", "You filled the whole board!")
    elif state["is_over"]:
        draw_message(ctx, width, height, "GAME OVER", "Press R or tap New to play again")
    elif state["is_paused"]:
        draw_message(ctx, width, height, "PAUSED", "Press P or tap Play to start")
